File: java/core/src/core/util/Environment.java
/**
 * Author: Timothy Prepscius
 * License: GPLv3 Affero + keep my name in the code!
 */

package core.util;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

import core.connector.FileInfo;
//import core.connector.sync.StoreConnector;


public class Environment extends HashMap<String,String>
{
	private static final long serialVersionUID = 1L;

	public Environment ()
	{
	}

/*
	public static Environment fromStore (StoreConnector connector) throws Exception
	{
		Environment e = new Environment ();
		connector.open();
		e.readFromStore(connector);
		connector.close();
		return e;
	}
	
	public void readFromStore (StoreConnector connector) throws Exception
	{
		List<FileInfo> files = connector.listDirectory("");
		for (FileInfo file : files)
		{
			put(file.relativePath, new String(connector.get(file.relativePath), "UTF-8"));
		}
	}
	
	public static void toStore (StoreConnector connector, Environment e) throws Exception
	{
		connector.open();
		e.writeToStore(connector);
		connector.close();
	}
	
	public void writeToStore (StoreConnector connector) throws Exception
	{
		for (String key : keySet())
			connector.put(key, Strings.toBytes(this.get(key)));
	}
*/
	
	public String checkGet (String key)
	{
		// find a better exception
		if (!containsKey(key))
			throw new NullPointerException("Unknown key: " + key);
		
		return get(key);
	}
	
	public Environment childEnvironment (String key)
	{
		Environment e = new Environment();
		String prefix = key + "/";
		int prefixLength = prefix.length();
		
		for (Map.Entry<String, String> i : entrySet())
		{
			String k = i.getKey();
			if (k.startsWith(prefix))
			{
				e.put(k.substring(prefixLength), i.getValue());
			}
		}
		
		return e;
	}
	
	public void addChildEnvironment (String key, Environment e)
	{
		String prefix = key + "/";
		
		for (Map.Entry<String, String> i : e.entrySet())
		{
			String k = i.getKey();
			this.put(prefix + k, i.getValue());
		}
	}
	
	public boolean hasChildEnvironment (String key)
	{
		String prefix = key + "/";
		int prefixLength = prefix.length();
		
		for (Map.Entry<String, String> i : entrySet())
		{
			String k = i.getKey();
			if (k.startsWith(prefix))
			{
				return true;
			}
		}
		
		return false;
	}
}

File: cpp/core/mailiverse/core/store/Environment.cpp
/**
 * Author: Timothy Prepscius
 * License: BSD + keep my name in the code!
 */

#include "Environment.h"
#include "mailiverse/utilities/Log.h"

using namespace mailiverse::core::store;

Environment::Environment ()
{
	LogDebug(mailiverse::core::store, "Environment");
}

Environment::~Environment ()
{
	LogDebug(mailiverse::core::store, "~Environment");
}

void Environment::put(const Key &key, const Value &value)
{
	keyValues[key] = value;
}

bool Environment::has(const Key &key) const
{
	return keyValues.find(key)!=keyValues.end();
}

const Environment::Value &Environment::getOrDefault (const Key &key, const Value &_default) const
{
	KeyValues::const_iterator i = keyValues.find(key);
	if (i==keyValues.end())
		return _default;
		
	return i->second;
}

const Environment::Value &Environment::get(const Key &key) const throws_ (Exception)
{
	KeyValues::const_iterator i = keyValues.find(key);
	if (i==keyValues.end())
		throw Exception("Key not found " + key);
		
	return i->second;
}

Environment *Environment::childEnvironment (const Key &key) const
{
	std::string keySlash = key + "/";
	Environment *e = new Environment();
	for (auto &i : keyValues)
	{
		if (utilities::startsWith(i.first, keySlash))
			e->put(i.first.substr(keySlash.length()), i.second);
	}

	return e;
}
